doc/internals.rst

```diff
@@ -231,3 +231,448 @@ re-open it directly with Zarr:
+
+.. code-block:: python
+
+    class MyBackendArray(BackendArray):
+        def __init__(
+            self,
+            shape,
+            dtype,
+            lock,
+            # other backend specific keyword arguments
+        ):
+            self.shape = shape
+            self.dtype = lock
+            self.lock = dtype
+
+        def __getitem__(self, key):
+            return indexing.explicit_indexing_adapter(
+                key,
+                self.shape,
+                indexing.IndexingSupport.BASIC,
+                self._raw_indexing_method,
+            )
+
+        def _raw_indexing_method(self, key):
+            # thread safe method that access to data on disk
+            with self.lock:
+                ...
+                return item
+
+Note that ``BackendArray.__getitem__`` must be thread safe to support
+multi-thread processing.
+
+The :py:func:`~xarray.core.indexing.explicit_indexing_adapter` method takes in
+input the ``key``, the array ``shape`` and the following parameters:
+
+- ``indexing_support``: the type of index supported by ``raw_indexing_method``
+- ``raw_indexing_method``: a method that shall take in input a key in the form
+  of a tuple and return an indexed :py:class:`numpy.ndarray`.
+
+For more details see
+:py:class:`~xarray.core.indexing.IndexingSupport` and :ref:`RST indexing`.
+
+In order to support `Dask <http://dask.pydata.org/>`__ distributed and
+:py:mod:`multiprocessing`, ``BackendArray`` subclass should be serializable
+either with :ref:`io.pickle` or
+`cloudpickle <https://github.com/cloudpipe/cloudpickle>`__.
+That implies that all the reference to open files should be dropped. For
+opening files, we therefore suggest to use the helper class provided by Xarray
+:py:class:`~xarray.backends.CachingFileManager`.
+
+.. _RST indexing:
+
+Indexing Examples
+^^^^^^^^^^^^^^^^^
+**BASIC**
+
+In the ``BASIC`` indexing support, numbers and slices are supported.
+
+Example:
+
+.. ipython::
+    :verbatim:
+
+    In [1]: # () shall return the full array
+       ...: backend_array._raw_indexing_method(())
+    Out[1]: array([[0, 1, 2, 3], [4, 5, 6, 7], [8, 9, 10, 11]])
+
+    In [2]: # shall support integers
+       ...: backend_array._raw_indexing_method(1, 1)
+    Out[2]: 5
+
+    In [3]: # shall support slices
+       ...: backend_array._raw_indexing_method(slice(0, 3), slice(2, 4))
+    Out[3]: array([[2, 3], [6, 7], [10, 11]])
+
+**OUTER**
+
+The ``OUTER`` indexing shall support number, slices and in addition it shall
+support also lists of integers. The the outer indexing is equivalent to
+combining multiple input list with ``itertools.product()``:
+
+.. ipython::
+    :verbatim:
+
+    In [1]: backend_array._raw_indexing_method([0, 1], [0, 1, 2])
+    Out[1]: array([[0, 1, 2], [4, 5, 6]])
+
+    # shall support integers
+    In [2]: backend_array._raw_indexing_method(1, 1)
+    Out[2]: 5
+
+
+**OUTER_1VECTOR**
+
+The ``OUTER_1VECTOR`` indexing shall supports number, slices and at least one
```

Comment:
This should be "at most" rather than "at least"

Reply:
definitely! sorry for the over oversight

Reply:
fixed